providers/nmea0183-signalk.js

```diff
@@ -40,52 +40,56 @@ function nmea0183ToSignalK (options) {
 
   this.parser = new Parser(options)
 
-  this.parser.on('nmea0183', sentence => {
-    options.app.signalk.emit('nmea0183', sentence)
-  })
+  // Object on which to send 'sentence' events
+  this.sentenceEventEmitter = options.app.signalk
 
+  // Prepare a list of events to send for each sentence received
+  this.sentenceEvents = ['nmea0183']
   if (options.sentenceEvent) {
-    ;(Array.isArray(options.sentenceEvent)
-      ? options.sentenceEvent
-      : [options.sentenceEvent]
-    ).forEach(event => {
-      this.parser.on('nmea0183', sentence => options.app.emit(event, sentence))
-    })
-  }
-
-  this.parser.on('signalk:delta', delta => {
-    if (this.timestamp) {
-      delta.updates.forEach(update => {
-        update.timestamp = this.timestamp
-      })
+    if (Array.isArray(options.sentenceEvent)) {
+      this.sentenceEvents = this.sentenceEvents.concat(options.sentenceEvent)
+    } else {
+      this.sentenceEvents.push(options.sentenceEvent)
     }
-
-    this.push(delta)
-  })
-
-  this.parser.on('warning', warning => {
-    debug(`[warning] ${warning.message}`)
-  })
-
-  this.parser.on('error', error => {
-    debug(`[error] ${error.message}`)
-  })
+  }
 }
 
 require('util').inherits(nmea0183ToSignalK, Transform)
 
 nmea0183ToSignalK.prototype._transform = function (chunk, encoding, done) {
-  if (Buffer.isBuffer(chunk)) {
-    chunk = chunk.toString().trim()
-  }
+  let sentence
+  let timestamp = null
 
   if (chunk && typeof chunk === 'object' && typeof chunk.line === 'string') {
-    this.timestamp = new Date(Number(chunk.timestamp))
-    this.parser.parse(`${chunk.line.trim()}\n`).catch(() => {})
+    timestamp = new Date(Number(chunk.timestamp))
+    sentence = chunk.line.trim()
```

Comment:
The old parser did internal line parsing. This setup assumes that the provider input is lines. I wonder if this is enough and the stuff out there still works.

I think for example https://github.com/SignalK/signalk-server-node/blob/master/settings/volare-tcp-settings.json might not work - it may work, if the source output is flushed and a tcp packet sent after every sentence, but that is incidental, not systematic. Like if you create a tcp server that produces a continuous stream of nmea0183 data will things still work? The official delimiter for nmea0183 is \r\n.

Reply:
> The old parser did internal line parsing.

What do you mean? It broke down blocks of multiple lines into individual lines? I am not sure it did: https://github.com/SignalK/signalk-parser-nmea0183/blob/07172ad569d5116c3262d9062651ba95afae2a7f/lib/index.js

I just created a config which connects to `localhost:1234` via TCP and then used `nc -l 1234` to paste blocks of 6 lines of NMEA data roughly once per second. The console shows 6 deltas/seconds and the log file includes all the lines that I have pushed. Once per line as expected.

I think in my case it works because I used the UI to create the connection and this generates the following config:
```
      "pipeElements": [
        {
          "type": "providers/simple",
          "options": {
            "logging": true,
            "type": "NMEA0183",
            "subOptions": {
              "validateChecksum": true,
              "type": "tcp",
              "host": "localhost",
              "port": "1234"
            }
          }
        }
      ],
      "id": "nmea-input-manual",
      "enabled": true
```

Looking at the source code of [`provider/simple.js`](https://github.com/SignalK/signalk-server-node/blob/master/providers/simple.js#L179) you can see it instantiates the `liner` transformer to break down lines and therefore everything works as expected.

However in the `volare-tcp-settings.json` you pointed to, the `tcp` provider is used which does not automatically include `liner`. Maybe the bug here is that this config file should include the `liner` transformer?

Reply:
Oh, my mistake: I was thinking of the ancient pre 2.0 version.

Now that we have the nice UI and `simple` provider I think we should remove the seriously outdated example setting file.